pkg/pillar/cmd/zedagent/localinfo.go

```diff
@@ -192,81 +203,199 @@ func processReceivedAppCommands(ctx *getconfigContext, cmdList *profile.LocalApp
+		if !hasLocalCmd {
+			appCmd = &types.LocalAppCommand{}
+			if ctx.localCommands.AppCommands == nil {
+				ctx.localCommands.AppCommands = make(map[string]*types.LocalAppCommand)
 			}
-			appCmd.Command = command
-			appCmd.LocalServerTimestamp = appCmdReq.Timestamp
-			appCmd.DeviceTimestamp = time.Now()
-			appCmd.Completed = false
-			cmdChanges = true
+			ctx.localCommands.AppCommands[appUUID.String()] = appCmd
+		}
+		if appCmd.Command == command &&
+			appCmd.LocalServerTimestamp == appCmdReq.Timestamp {
+			// already accepted
 			continue
 		}
-		// Add new entry.
-		ctx.localAppCommands.Cmds = append(ctx.localAppCommands.Cmds,
-			types.LocalAppCommand{
-				AppUUID:              appUUID,
-				Command:              command,
-				LocalServerTimestamp: appCmdReq.Timestamp,
-				DeviceTimestamp:      time.Now(),
-				Completed:            false,
-			})
+		appCmd.Command = command
+		appCmd.LocalServerTimestamp = appCmdReq.Timestamp
+		appCmd.DeviceTimestamp = time.Now()
+		appCmd.Completed = false
+
+		appCounters, hasCounters := ctx.localCommands.AppCounters[appUUID.String()]
+		if !hasCounters {
+			appCounters = &types.LocalAppCounters{}
+			if ctx.localCommands.AppCounters == nil {
+				ctx.localCommands.AppCounters = make(map[string]*types.LocalAppCounters)
+			}
+			ctx.localCommands.AppCounters[appUUID.String()] = appCounters
+		}
+
+		// Update configuration to trigger the operation.
+		switch appCmd.Command {
+		case types.AppCommandRestart:
+			appCounters.RestartCmd.Counter++
+			appCounters.RestartCmd.ApplyTime = appCmd.DeviceTimestamp.String()
+			appInst.LocalRestartCmd = appCounters.RestartCmd
+			publishApps[appUUID.String()] = *appInst
+
+		case types.AppCommandPurge:
+			appCounters.PurgeCmd.Counter++
+			appCounters.PurgeCmd.ApplyTime = appCmd.DeviceTimestamp.String()
+			appInst.LocalPurgeCmd = appCounters.PurgeCmd
+			// Trigger purge of all volumes used by the application.
+			// XXX Currently the assumption is that every volume instance is used
+			//     by at most one application.
+			if ctx.localCommands.VolumeGenCounters == nil {
+				ctx.localCommands.VolumeGenCounters = make(map[string]int64)
+			}
+			for i := range appInst.VolumeRefConfigList {
+				vr := &appInst.VolumeRefConfigList[i]
+				uuid := vr.VolumeID.String()
+				remoteGenCounter := vr.GenerationCounter
+				localGenCounter := ctx.localCommands.VolumeGenCounters[uuid]
+				// Un-publish volume with the current counters.
+				volKey := volumeKey(uuid, remoteGenCounter, localGenCounter)
+				volObj, _ := ctx.pubVolumeConfig.Get(volKey)
+				if volObj == nil {
+					log.Warnf("Failed to find volume %s referenced by app instance "+
+						"with UUID=%s - not purging this volume", volKey, appUUID)
+					continue
+				}
+				volume := volObj.(types.VolumeConfig)
+				unpublishVols = append(unpublishVols, volKey)
+				// Publish volume with an increased local generation counter.
+				localGenCounter++
```

Comment:
If an app instance has multiple volumes you increase this for each one. That seems unnecessary. 
Shouldn't you instread just increment (and set to 1 if doesn't exist) ctx.localCommands.VolumeGenCounters[uuid]? Or pick a localGenCounter value for the app instance and add that for all of the volumes?

Reply:
OK